cpp/src/arrow/array/array_base.h

```diff
@@ -207,6 +207,14 @@ static inline std::ostream& operator<<(std::ostream& os, const Array& x) {
 
 /// Base class for non-nested arrays
 class ARROW_EXPORT FlatArray : public Array {
+ public:
+  /// Returns the slice buffer for the second buffer in the
+  /// array (values buffer for primitives, offsets buffers
+  /// for variable length binary types.
+  ///
+  /// \note API EXPERIMENTAL
+  virtual Result<std::shared_ptr<Buffer>> SlicedValues(MemoryPool* pool) const = 0;
```

Comment:
Since this is only used in one place and uses "Slice" with very different semantics from other slices, I think it'd be better to
1. move it into `column_writer.cc` or
2. provide a more encapsulated and easily understood function, for example `Result<std::shared_ptr<ArrayData>> arrow::internal::ReplaceNullBitmap(const ArrayData& data, std::shared_ptr<Buffer> null_bitmap, MemoryPool* pool)`